text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,214 @@
+
+Conversely, if a script running in `virtual-script` needs access to a global variable from the main window, we can _import_ this global. 
+
+A syntax like the following will be provided:
+
+```html
+<virtual-script exported-globals="foo"></virtual-script>
+<virtual-script imported-globals="bar"></virtual-script>
+```
+
+- `virtual-script` container would not have access to globals defined by other scripts in the outer window unless they were imported using `imported-globals`.
+
+- Globals defined in `exported-globals` will be redefined in any `virtual-script` container.
+- Exported globals will also be available to any Locker-created sandboxes
+- Globals defined in Oasis using `imported-globals` will be redefined only in the `virtual-script` container they are imported to.
+
+Note: these global values are controlled side-channels between the main window and the sandbox, and should be used with caution. For example, creating a component that depends on `ga` to be defined, is an anti-pattern.
+
+### Supported Attributes
+
+`crossorigin`: indicates that the browser should provide more information to window.onerror when handling CORS errors
+
+`defer`: indicates that the script should be executed after the document has parsed (has no effect if `src` is absent)
+
+`integrity`: enables the browser to verify that a fetched resource is delivered without unexpected manipulation
+
+`nomodule`: the script should not be executed in browsers that support ES2015 modules
+
+`nonce`: user provided nonce values for CSP will be supported
+
+`referrerpolicy`: indicates which referrer to send when fetching the script
+
+`src`: the address of the resource
+
+**async**
+
+TBD describe asynchronous nature of `virtual-script`.
+
+### Unsupported Attributes
+
+`type`: all `virtual-script` resources will be `text/javascript`. `type="module"` will not be supported.
+
+`charset`: this attribute is deprecated
+
+`language`: this attribute is deprecated
+
+`nonce`: has special semantics since it is a global attribute and browsers do special behavior with this attribute when a CSP policy is applied. We will have to treat this attribute specially.
+
+
+### Known Limitations
+
+* Currently there is an outstanding issue with Synthetic Shadow in that the Event object patching happens lazily, so we are unable to prevent its patching at the time of creating the sandbox, which means that events are still re-targeted in some cases. [See LWC PR #1569](https://github.com/salesforce/lwc/pull/1569).
+
+### Testing
+
+We will want to gather the top 20 common libraries and implement their behavior inside of `virtual-script` to ensure that we are properly accounting for all these libraries' use cases. This will be our integration test suite.
+
+## Drawbacks
+
+- The element may end up being anywhere from 5kb - 20kb.
```

Comment:
Is this something that can be extracted as a plugin? Seems unfortunate that I'd have to incur 5-20kb and all the complexities it brings (patching native methods) for a feature that seems intended for platform?

Reply:
This feature will not be a part of core LWC. In fact it's intended to be used only in our b2c products today and will be a separate library.